### ISSUE TITLE
No link to view details after updating

### ISSUE THREAD

**Issue by johnalarcon** [codepotent/Update-Manager#3]
### Feature Description
The list table in the plugin admin page does not show a link to "View Details" after applying an update. The link shows before the update, just, not after. Such a link isn't a requirement, but, has come to be fairly expected and would be handy for viewing the plugin details.

### Context
This link will give users more information about the plugins on which they rely.

H/T: [xxsimoxx](https://github.com/xxsimoxx)

**Comment by johnalarcon** [comment]
@xxsimonxx, following up from [here](https://forums.classicpress.net/t/classicpress-update-manager/1707/28).

Were you saying:

1) the Update Manager plugin needs a link to view its details in its own row, _or_
2) the remotely-updatable plugin needs a link to view its details in its own row, _or_
3) both of the above?

If you were only referring to **1**, then my questions are answered. If you were referring to **2** or **3**, please consider the following.

All plugins _should_ have a link to view details in their own row. No doubt about it. The question here is: _is it the job of the Update Manager to create this link for other plugins_? I'm not sure it is...but, I'm also not sure it isn't. Creating this link would usually be left to the plugin developer, but, given that the Update Manager plugin does indeed interact with and present these details, it does _somewhat_ make sense to add the link for the plugins connected to the Update Manager.

To clarify, it's no more or less work to implement it one way or the other; I've already built the solution and just await your input before deciding which file it lives in.

**Comment by xxsimoxx** [comment]
I was only referring to 1.

The link is created (as I can see) from CP/WP when it can find more details on the wp.org directory.
It's not the plugin itself adding *View Details* (you can see this in hello dolly).

GitHub updater adds the link when it can provide more details, but it doesn't seem a good example, because it's done forcedly.

I can be totally wrong, but it seems that when Update Manager tells CP that there is an update, it also tells other thing about the plugin, and then *View Details* is displayed, but not pass those info when there is no update available.

**Comment by johnalarcon** [comment]
Thanks for clarifying.

In terms of the Update Manager plugin, it sets a flag that indicates the plugin is external. Core doesn't show the link for plugins that are external. Removing [that flag](https://github.com/codepotent/Update-Manager/blob/d790e2a6fb28ea80c940c0e451d6ce5bea80a8fa/includes/functions.php#L411) and amending [the related check](https://github.com/codepotent/Update-Manager/blob/d790e2a6fb28ea80c940c0e451d6ce5bea80a8fa/classes/UpdateClient.class.php#L286) for it causes the system to not realize it's an externally-hosted plugin and the link appears automatically. My guess is that hiding the link for external plugins is a preventative measure of sorts since the popular updater plugins don't pull in enough data to populate the modal windows and it sort of just looks broken (even though it technically still works.) ..."improved UX" we'll call it. :) The Update Manager plugin does account for all the fields and fully populates the modal, so, perhaps I'll look into removing the flag rather than adding an additional method to add the link.

**Comment by xxsimoxx** [comment]
I've checked [where CP chooses](https://github.com/ClassicPress/ClassicPress/blob/a76cb4ae8867a1c069646e8f33ed91e36b5ca642/src/wp-admin/includes/class-wp-plugins-list-table.php#L776) to show the *view details* link.
It seems that the **slug** must be defined, and it's not at the moment of displaying plugins.

I've also tried to remove the flag you referenced above (and the related check) but nothing changed. 
Hope it helps...

**Comment by xxsimoxx** [comment]
Maybe I'm wrong, but `filter_plugins_api_result` never get called!
I added a `trigger_error('I'm there!', E_USER_WARNING);` but nothing appears in log.

**Comment by johnalarcon** [comment]
I've been looking at this further and it has proven to be a tougher issue than I thought. I'm not entirely sure what is the best way to resolve it.

For example, let's say I pushed an update for the Update Manager plugin in Pending status. If you were whitelisted, you'd get the endpoint data and the modal would populate as expected. If you weren't whitelisted, you'd get the error screen in the modal.

_The thing is_: there needs to be a way to show _at least some_ data regardless of the state of the remote endpoint.

This makes me think that saving the endpoint data (also) in the options table of the end-user might be a solution. Like, if they check for updates and the endpoint is empty (pending, draft, trashed, site down) then, the last-saved data could be pulled from their own options table instead. To ensure that data remained fresh, the option data could be refreshed every time they checked for an update while the endpoint was in _publish_ status. Just an idea. Like I said, I'm not sure what is the best solution. 

I'm happy to hear any feedback or alternate ideas.

**Comment by xxsimoxx** [comment]
I like this way.

Maybe a transient with long expiration. This way the plugin developer have not to take care of deleting it on unistall. Say `30 * DAYSINSECONDS` could leave room for a "lazy" testing.

Maybe in the form `__NAMESPACE__ . $version` so if the user is playing with different versions and the transient is not set we could do something...

**Comment by johnalarcon** [comment]
A bit misleadingly, a transient's expiration is actually the _maximum life_ of the transient; it's not an actual guarantee that the data will persist for that amount of time. Transients can be deleted at any time either manually or by so-called `database cleanup` or `transient management` plugins – they aren't recommended for persisting data more than a few minutes. But, I do like the idea of not adding more uninstall work for developers...

Hmm.

I think _most likely_, the (last-updated) data probably won't ever even be needed...but, I don't think that we can safely write the code under that assumption without leaving the error in place for a few edge cases.

Maybe a good compromise would be to store the data in a transient, as you suggested, and simply hide the link to view more details if that transient isn't present. This would prevent users from clicking a link that produces an error. However, this would create a different issue: end users wouldn't actually see the link to view more details until the next time their site checked for updates. A workaround for _that_ issue would be to force the system to check for updates when the plugin is activated.

I'm not sure which way creates a better UX. If the update check is forced, it will make plugin activation take longer...but, if the link is left out, the user may feel they can't get valuable information they need right after activating the plugin. Do you have an opinion on which way would create a better UX?

**Comment by xxsimoxx** [comment]
Really these are edge cases.
If I run a plugin that erases "unnecessary" data I would not scare something strange.
Hiding the link (and maybe restore the link to plugin site) seems to me the best for UX.
And maybe at that time (when we found the transient has gone) force the system to check for updates.

**Comment by johnalarcon** [comment]
Sorry for the delay in getting out an RC2. I was hoping for sometime during the week and here it is the end of the weekend already. On that note, I've worked out an elegant solution using some of what we've discussed with a slightly different approach to make it easier. 💡

The "View Details" link now displays and populates the remote modal windows without any special requirements. If you have (at least) the minimum required fields filled out in the endpoint editor and have hit _Publish_ at least once on it, your remote users will enjoy populated modals whether there is an update available or not. The solution doesn't require storage or transients on the remote site, there's no forced update checks to initialize the data, and there's no need to hide the link for edge cases. It just works. :) I think RC2 will be out in the next day or so.
